이미지 업로더 구현

makepin2r opened the pull request:
# 구현 기능
## UI 구현
- [x] 파일 업로드 UI
  - [x] 일반 파일 선택
  - [x] drag & drop 
- [x] 업로드 진행률 프로그레스 UI
- [x] 업로드된 이미지 파일 리스트 UI
  - [x] 파일명, 업로드 일시, 용량, 썸네일 추가
- [x] 미리보기 UI
  - [x] 미리보기 취소 버튼 추가
- [x] 업로드 전 확인 모달

- [x] `(optional)` 업로드 목록 (파일명, 진행률, 중단 버튼)

## 기능 구현
- [x] 파일 업로드
  - [x] 파일 선택창에서 추가
  - [x] drag & drop으로 추가
  - [x] 업로드 진행률 표시 ([8630ed6](https://github.com/f-lab-edu/image-uploader/pull/1/commits/8630ed6066eb46f7854d34fdca930912284ff70f), [862e4ca](https://github.com/f-lab-edu/image-uploader/pull/1/commits/862e4ca36f37427c88eaac34f2601e938d72b373))
  - [x] 확장자 validation
  - [x] 여러 개 업로드 시 허용 확장자 파일만 업로드 가능 ([cf60e73](https://github.com/f-lab-edu/image-uploader/pull/1/commits/cf60e73b6925b71182b87adf6cd956946efad585))
- [x] 업로드 된 파일 리스트로 출력
- [x] 리스트에서 선택한 파일 미리보기 ([eb5a8ad](https://github.com/f-lab-edu/image-uploader/pull/1/commits/eb5a8ad8efd4436783e084aa98e902fc01cd359e), [fe76fdc](https://github.com/f-lab-edu/image-uploader/pull/1/commits/fe76fdca31cf93eff87c0ffbf47c97d38209f101))

(25. 08. 05)
피드백 이후 추가한 기능은 위 '기능 구현' 항목에 커밋 링크 추가해뒀습니다!

makepin2r committed:
feat: 이미지 업로드 UI 컴포넌트 정의

makepin2r committed:
update: 이미지 업로드 UI 드래그 이펙트 추가

makepin2r committed:
feat: 이미지 미리보기 UI 정의

makepin2r committed:
update: 이미지 미리보기 UI prop 정의

makepin2r committed:
feat: 업로드된 이미지 리스트 UI 컴포넌트 정의

makepin2r committed:
fix: 업로드된 이미지 선택 UI 렌더링 시 key 추가

makepin2r committed:
update: 이미지 미리보기 UI 컴포넌트 예외 처리 추가
- image props 타입 null 포함
- image props 변경 시 image가 null일 경우 예외 처리 추가
- image null일 경우 placeholder 추가

makepin2r committed:
update: 이미지 업로더 UI 파일 업로드 기능 추가
- 파일 업로드 함수 props 추가
- 파일 선택 추가 시 / 드래그로 추가 시 파일 업로드 로직 연결

makepin2r committed:
update: 이미지 업로드 UI 확장자 밸리데이션 추가

makepin2r committed:
chore: 이미지 업로드 UI 안내 문구 수정
- 오류 alert에 제한된 확장자 표시

makepin2r committed:
chore: App.tsx 업데이트
- 실제 사용 예시 코드 작성

makepin2r committed:
delete: 더미 이미지 삭제

f-lab-stella commented on src/App.tsx:
```diff
@@ -1,7 +1,36 @@
+      <div className="flex flex-col gap-4">
+        <ImageUploader
+          onUploadImages={handleUploadImages}
+          acceptedExt={["jpg", "jpeg"]}
```
해당필드가 동작하는지 다른 확장자 업로드 해보시겠어요?! 미동작한다면 js에서 확장자 검사가 필요해보입니다~

f-lab-stella commented on src/components/image-preview.tsx:
```diff
@@ -0,0 +1,47 @@
+}
+
+const ImagePreview = ({ image }: ImagePreviewProps) => {
+  const [previewImg, setPreviewImg] = useState<string>("");
```
`previewImagUrl`이 어떨까요??

makepin2r replied:
[99d4ae1](https://github.com/f-lab-edu/image-uploader/pull/1/commits/99d4ae1fad9eeb4a222206bfbb449bf237220eef) 에서 수정했습니다.

f-lab-stella commented on src/components/image-preview.tsx:
```diff
@@ -0,0 +1,47 @@
+          alert("이미지를 읽는 데 실패했습니다.");
+          console.warn("unexpected type from FileReader:", reader.result);
+          setPreviewImg("");
+        }
```
이부분 try/catch로도 가능할까요~??

makepin2r replied:
outDate되어서 수정한 곳에 코멘트 추가하겠습니다!

makepin2r replied:
reader.onloadend 내의 로직을 조건문이 아닌 try-catch 문으로 처리하면 좋겠다는 말씀이 맞을까요??

f-lab-stella replied:
1. onloadend 부분만 감싸는 방법도 있습니다.
2. else 부분을 try/catch로 감싸는 방법도 있습니다.

여기서 제가 리뷰드린건 1번이었어용~

f-lab-stella commented on src/components/image-uploader.tsx:
```diff
@@ -0,0 +1,117 @@
+    if (!validateUploadedImages(targetFiles)) {
+      alert(`${acceptedExt.join(", ")} 확장자의 파일만 추가할 수 있습니다.`);
+      return;
+    }
```
여기서 확장자 검사하고있군요 👍

f-lab-stella commented:
[기존기능에 대한 피드백]
1. 여러개 업로드시, 허용되는 확장자는 업로드 되어야 합니다.
2. 이미지는 업로드 되면 바로 리스트에서 볼 수 있어야합니다.
3. 이미지 미리보기에서 x누르면 placeholder로 변경되어야합니다.

[추가기능]
1. 이미지 용량제한
2. 업로드 중 특정 파일 업로드 중단 (옵셔널)
3. 이미지 목록에서 이미지 삭제 기능
4. 중복이미지 핸들링
- 업로드된다.
- 목록에서 각 파일의 파일명 / 업로드시간(YYYY.MM.DD.HH) / 용량 표기

f-lab-stella commented on src/components/image-uploader.tsx:
```diff
@@ -0,0 +1,117 @@
+  const [isDragOver, setIsDragOver] = useState<boolean>(false);
+  const isActive = isHover || isDragOver;
+
+  const handleDrag = (
+    e: React.DragEvent<HTMLLabelElement>,
+    status: boolean
+  ) => {
+    e.preventDefault();
+    setIsDragOver(status);
+  };
+
+  const handleHover = (
+    e: React.MouseEvent<HTMLLabelElement>,
+    status: boolean
+  ) => {
+    e.preventDefault();
+    setIsHover(status);
+  };
```
preventDefault제거 가능한지 확인부탁드려요~

makepin2r replied:
[65db170](https://github.com/f-lab-edu/image-uploader/pull/1/commits/65db170ecca74f02972962e101bfe36c203c1b9c) 에서 수정하였습니다.
확인해보니 drag 이벤트를 제외하곤 모두 preventDefault 제어가 필요 없어 보였어요!

드래그 이벤트에서 preventDefault를 생략했을 때는 
- 드래그한 상태로 커서가 영역에서 벗어나도 drag state가 유지되거나, 
- 해당 영역에 drop했을 경우 제가 작성한 로직이 아닌 원본 리소스가 브라우저에서 오픈되는 등의 사이드 이펙트가 있었습니다.

찾아보니 드래그 이벤트느 브라우저에서 기본으로 지정된 이벤트 수행 동작이 존재하는 경우가 있어 기본 동작을 해제해주기 위해 preventDefault가 필요하다고 합니다. ([참고](https://web.dev/articles/drag-and-drop#file-upload))

makepin2r committed:
chore: 이미지 미리보기 UI 내 state 변수명 변경
- previewImg -> previewImgUrl

makepin2r committed:
update: 업로드된 이미지 리스트 UI 삭제 기능 추가
- 미리보기 선택 함수 인자 File -> 인덱스로 변경
- 삭제 함수 props 추가

makepin2r committed:
update: 업로드된 이미지 삭제 로직 추가
- 삭제 함수 추가 및 UI에 전달
- 미리보기 이미지 state File -> 인덱스로 변경

makepin2r committed:
chore: 테스트용 props 전달 삭제

makepin2r committed:
update: 미리보기 이미지 props 예외 처리 추가

makepin2r committed:
build: svgr 설치

makepin2r committed:
feat: 버튼 공통 컴포넌트 정의

makepin2r committed:
feat: 삭제 버튼 컴포넌트 정의

makepin2r committed:
update: 파일 업로드 방식 변경
- FileInfo 타입 정의 (url 및 기본 정보)
- 미리보기 이미지 컴포넌트 File url 변환 로직 삭제, props FileInfo로 변환
- 이미지 업로드 컴포넌트 File 변환 로직 추가

makepin2r committed:
update: 업로드 이미지 UI 수정
- 썸네일, 업로드 일시, 용량 정보 추가
- 데이터 타입 File -> FileInfo로 변경

makepin2r committed:
style: 스타일 수정
- 이미지 미리보기 캡션
- 공통 버튼 컴포넌트 기본 스타일

makepin2r committed:
delete: 불필요 기본 허용 확장자 삭제

makepin2r committed:
chore: DeleteIcon svg 리소스 이름 변경

makepin2r committed:
feat: refresh 아이콘 버튼 정의

makepin2r committed:
style: 버튼 공통 컴포넌트 스타일 추가
- 트랜지션

makepin2r committed:
update: 이미지 미리보기 컴포넌트 이미지 선택 취소 기능 추가

makepin2r committed:
update: 이미지 업로드 컴포넌트 허용된 확장자 파일만 업로드 처리 기능 추가
- validation 함수 변경 : 모든 업로드 파일 확장자 허용 여부 체크 -> 허용된 파일만 filter
- 미허용 확장자 파일이 존재할 경우 alert 처리

makepin2r committed:
style: 이미지 미리보기 UI 배경색 변경

makepin2r committed:
feat: 프로그레스 바 UI 정의

makepin2r committed:
feat: 이미지 업로드 상태창 UI 정의
- 컴포넌트 정의
- FileUploadInfo 타입 정의 (업로드 상태 정보)

makepin2r committed:
update: 이미지 업로드 컴포넌트 파일 업로드 상태창 기능 추가
- 파일 업로드 상태 표시 변수 추가
- FileReader 콜백 수정 (상태 표시 변수 업데이트)
- 상태창 UI 컴포넌트 연결

makepin2r committed:
fix: 이미지 업로드 상태 업데이트 시 잘못된 로직 수정 (readyState)

makepin2r committed:
chore: UI 확인용 코드 정리

makepin2r committed:
chore: 컴포넌트 이름 변경 (Button -> IconButton)

makepin2r committed:
chore: 아이콘 리소스 추가

makepin2r committed:
chore: 기본 이미지 허용 확장자 배열 컴포넌트와 분리

makepin2r committed:
feat: 모달 공통 컴포넌트 정의 (Dialog)

makepin2r committed:
feat: 이미지 업로드 전 확인 모달 정의

makepin2r committed:
update: 이미지 업로드 컴포넌트 기능 추가
- 이미지 업로드 확인 모달 추가
- 확인 모달 전달용 state 추가 (UI 제어, 데이터 전달용)
- 이미지 업로드 함수 분리
  - 업로드 확인 전 validation, FileList-> File 변환 로직 분리(handleImagesToConfirm)

makepin2r committed:
delete: 디버깅용 불필요 코드 삭제

makepin2r committed:
delete: 이미지 업로드 컴포넌트 내 불필요한 이벤트 버블링 취소 삭제

makepin2r committed:
update: 이미지 업로드 컴포넌트 용량 제한 로직 추가

makepin2r committed:
style: 이미지 업로드 확인 모달 미리보기 스타일링 수정 (썸네일 크기)

f-lab-stella commented on src/components/ui/confirm-image-upload-dialog.tsx:
```diff
@@ -0,0 +1,126 @@
+  };
+
+  const selectFileToConfirm = (i: number) => {
+    let newSelectedFiles = [];
+    if (selectedFiles.includes(i)) {
+      newSelectedFiles = selectedFiles.filter((value) => value !== i);
+    } else {
+      newSelectedFiles = [...selectedFiles, i];
+    }
```
```suggestion
    const newSelectedFiles = selectedFiles.includes(i) ? selectedFiles.filter((value) => value !== i) : [...selectedFiles, i]
```

makepin2r replied:
[4b1cd56](https://github.com/f-lab-edu/image-uploader/pull/1/commits/4b1cd5636f02c4aaeecf551499f5b5bfc750691e) 에서 수정했습니다.

f-lab-stella reviewed (commented):
`sleep`은 어떻게 추가된 함수일까요 ~?

makepin2r commented:
[여기 코멘트](https://github.com/f-lab-edu/image-uploader/pull/1#pullrequestreview-3091949250)에 왜인지 답글창이 안뜨네요..?

제가 부연 설명을 깜박했습니다!
sleep은 실제 로직이랑은 무관합니다 🥲
파일이 너무 순식간에 업로드가 되어버려서, 파일 업로드 진행률 보여주는 UI가 작성 의도대로 동작하는지 확인하기 위해 넣은 코드였습니다.
혹시 멘토님도 실행해보실 수 있을 것 같아 일단 남겨뒀어요! 
이후에는 지우겠습니다.

makepin2r committed:
refactor: 이미지 미리보기 모달 파일 선택 로직 리팩토링

makepin2r committed:
update: 이미지 업로드, 미리보기 시 DataURL 변환 로직 에러 핸들링 수정
- try-catch 문으로 런타임 에러 대응

f-lab-stella commented:
SonarCloud에서 중복코드가 많다고 알려주네요 이 부분 확인해보면 좋겠습니다.
https://sonarcloud.io/component_measures?id=f-lab-edu_image-uploader&pullRequest=1&metric=new_duplicated_lines_density&view=list

- 구두로 추후 공통함수로 빼는 것으로 이야기되었습니다.

f-lab-stella commented on src/components/ui/confirm-image-upload-dialog.tsx:
```diff
@@ -27,15 +27,25 @@ const ConfirmImageUploadDialog = ({
+              createdAt: new Date(),
+            });
+          } else {
+            reject(new Error("Unexpected result type from FileReader."));
```
여기서는 throw를 날리면 되겠네요

makepin2r replied:
[3aaf658](https://github.com/f-lab-edu/image-uploader/pull/1/commits/3aaf658e45d984e2e474ee7c38619b73f62d2bd6)에서 함께 수정하였습니다.

f-lab-stella commented on src/components/image-uploader.tsx:
```diff
@@ -142,15 +142,25 @@ const ImageUploader = ({
-          } else {
-            reject(new Error("Unexpected result type from FileReader."));
+          try {
+            if (typeof reader.result === "string") {
```
await updateFileUploadStatus(event, file, i); 이 라인이 try안으로 들어와야되겠네요~

makepin2r replied:
말씀하신 코드는 UI 업데이트용 코드여서 그대로 두었습니다!

파일 처리 로직은 [3aaf658](https://github.com/f-lab-edu/image-uploader/pull/1/commits/3aaf658e45d984e2e474ee7c38619b73f62d2bd6)에서 공통 로직으로 변경되었습니다.

makepin2r committed:
refactor: FileReader onloadend 콜백 로직 공통화
- FileReader로 파일 처리 시 File -> FileInfo 변환하는 공통 로직 함수화

makepin2r committed:
update: 이미지 업로드 UI, 이미지 업로드 확인 모달 파일 처리 로직 수정
- onloadend 로직에 handleFileOnLoadEnd 적용

makepin2r committed:
chore: 이미지 업로드 시 파일 용량 validation 안내 문구 수정

makepin2r committed:
delete: 불필요 주석 삭제

makepin2r commented on src/utils/file-handler.ts:
```diff
@@ -0,0 +1,32 @@
+}: {
+  reader: FileReader;
+  file: File;
+  onResolve: (value: FileInfo | PromiseLike<FileInfo>) => void;
```
인자 타입에 대해서 찾아보았는데, new Promise()의 resolve가 받을 수 있는 인자의 타입으로
- 일반 값
- Promise 객체
- PromiseLike 객체 (then 메서드를 갖고 있는 유사 Promise 객체)
이 세 가지가 있다고 합니다. ([mdn 문서](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve))
그런데 타입스크립트 내에서 Promise는 PromiseLike를 상속하고 있어 `FileInfo | PromiseLike<FileInfo>` 라는 타입 추론이 나온 것 같아요!

resolve가 기대하는 인자에 대해 이 이상의 타입 추상화는 어려울 것 같지만, 인자 타입 자체를 일반 FileInfo로 고정하고 작성 시점에 전달하는 인자 타입을 좁히는 방법도 있을 것 같아요.
